process_report/invoices/discount_invoice.py

```diff
@@ -0,0 +1,79 @@
+from dataclasses import dataclass
+
+import pandas
+
+import process_report.invoices.invoice as invoice
+
+
+@dataclass
+class DiscountInvoice(invoice.Invoice):
+    """
+    Invoice class containing functions useful for applying discounts
+    on dataframes
+    """
+
+    @staticmethod
+    def apply_flat_discount(
+        invoice: pandas.DataFrame,
+        pi_projects: pandas.DataFrame,
+        discount_amount: int,
+        discount_field: str,
+        balance_field: str,
+        code_field: str = None,
+        discount_code: str = None,
+    ):
+        """
+        Takes in an invoice and a list of PI projects that are a subset of it,
+        and applies a flat discount to those PI projects. Note that this function
+        will change the provided `invoice` Dataframe directly. Therefore, it does
+        not return the changed invoice.
+
+        This function assumes that the balance field shows the remaining cost of the project.
+        Basically, the provided invoice must show that balance = cost.
```

Comment:
Balance field showing the remaining cost of a project is not the same as balance = cost. In the case of a previous invoice having applied a discount, balance would be different from cost, and that is ok. 

You bring a good point though with regards to Balance having to be preinitialized before calling apply flat discount.

Reply:
I should have clarified that the balance field should equal **what the PI would pay if the flat discount is not applied**, not just cost. I'll clarify this in the docstring.